contracts/beacon/Voucher.sol

```diff
@@ -0,0 +1,469 @@
+// SPDX-FileCopyrightText: 2024 IEXEC BLOCKCHAIN TECH <contact@iex.ec>
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.20;
+
+import {IexecLibCore_v5} from "@iexec/poco/contracts/libs/IexecLibCore_v5.sol";
+import {IexecLibOrders_v5} from "@iexec/poco/contracts/libs/IexecLibOrders_v5.sol";
+import {IexecPoco1} from "@iexec/poco/contracts/modules/interfaces/IexecPoco1.v8.sol";
+import {IexecPoco2} from "@iexec/poco/contracts/modules/interfaces/IexecPoco2.v8.sol";
+import {IexecPocoAccessors} from "@iexec/poco/contracts/modules/interfaces/IexecPocoAccessors.sol";
+import {IexecPocoBoost} from "@iexec/poco/contracts/modules/interfaces/IexecPocoBoost.sol";
+import {IexecPocoBoostAccessors} from "@iexec/poco/contracts/modules/interfaces/IexecPocoBoostAccessors.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {IVoucherHub} from "../IVoucherHub.sol";
+import {IVoucher} from "./IVoucher.sol";
+
+/**
+ * @title Implementation of the voucher contract.
+ * Note:
+ *  - This contract and the Beacon are deployed using "Upgrades" plugin of OZ.
+ *  - Vouchers ownership must not be transferable.
+ */
+contract Voucher is Initializable, IVoucher {
+    // keccak256(abi.encode(uint256(keccak256("iexec.voucher.storage.Voucher")) - 1))
+    // & ~bytes32(uint256(0xff));
+    bytes32 private constant VOUCHER_STORAGE_LOCATION =
+        0xc2e244293dc04d6c7fa946e063317ff8e6770fd48cbaff411a60f1efc8a7e800;
+
+    /// @custom:storage-location erc7201:iexec.voucher.storage.Voucher
+    struct VoucherStorage {
+        address _owner;
+        address _voucherHub;
```

Comment:
If it becomes immutable, how to update storage slot if someday VoucherHub address changes? Should we consider we are safe since VoucherHub is UUPSUpgradeable